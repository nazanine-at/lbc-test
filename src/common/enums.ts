export enum EMessageType {
  PUBLIC = 'PUBLIC',
  PRIVATE = 'PRIVATE',
}
