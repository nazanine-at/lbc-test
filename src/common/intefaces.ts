import { EMessageType } from './enums'

export interface IState {
  messages: IMessage[]
  selectedMessage: IMessage
}

export interface IMessage {
  userId?: string
  subject?: string
  message?: string
  viewType?: EMessageType
}
