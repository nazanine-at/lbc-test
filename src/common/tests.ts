import configureMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store'
import { IState } from './intefaces'

const initStore: MockStoreCreator = configureMockStore()
export function mockStore(initialState: IState): MockStoreEnhanced {
  return initStore(initialState)
}
