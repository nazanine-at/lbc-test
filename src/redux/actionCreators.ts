import { IMessage } from '../common/intefaces'

// export const setMessagesCreator = (messages: IMessage[]) => {
//   return { type: 'messages/setMessages', payload: messages }
// }
//
// export const addMessageCreator = (message: IMessage) => {
//   return {
//     type: 'message/addMessage',
//     payload: message,
//   }
// }
//
// export const selectMessageCreator = (message: IMessage | undefined) => {
//   return { type: 'message/selectMessage', payload: message }
// }
//
// export const unselectMessageCreator = () => {
//   return { type: 'message/unselectMessage', payload: {} }
// }
