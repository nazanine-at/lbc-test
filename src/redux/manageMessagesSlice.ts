/**
 * This is a reducer - a function that takes a current state value and an
 * action object describing "what happened", and returns a new state value.
 * A reducer's function signature is: (state, action) => newState
 *
 * The Redux state should contain only plain JS objects, arrays, and primitives.
 * The root state value is usually an object.  It's important that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * You can use any conditional logic you want in a reducer. In this example,
 * we use a switch statement, but it's not required.
 */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { IState } from '../common/intefaces'

export const fetchMessages = createAsyncThunk('fetchMessages', async () => {
  const res = await fetch('/messages')
  return res.json()
})

export const initialState: IState = {
  messages: [],
  selectedMessage: {},
}

const manageMessagesSlice = createSlice({
  name: 'manageMessages',
  initialState,
  reducers: {
    addMessage: (state: IState, { payload }): void => {
      state.messages = [...state.messages, payload]
    },
    selectMessage: (state: IState, { payload }): void => {
      state.selectedMessage = payload
    },
    unselectMessage: (state: IState): void => {
      state.selectedMessage = {}
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMessages.fulfilled, (state, { payload }) => {
      state.messages = payload
    })
    builder.addCase(fetchMessages.rejected, (state) => {
      state.messages = []
    })
  },
})

export const { unselectMessage, selectMessage, addMessage } = manageMessagesSlice.actions
export default manageMessagesSlice.reducer
