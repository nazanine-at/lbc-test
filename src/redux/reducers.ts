/**
 * This is a reducer - a function that takes a current state value and an
 * action object describing "what happened", and returns a new state value.
 * A reducer's function signature is: (state, action) => newState
 *
 * The Redux state should contain only plain JS objects, arrays, and primitives.
 * The root state value is usually an object.  It's important that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * You can use any conditional logic you want in a reducer. In this example,
 * we use a switch statement, but it's not required.
 */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { IState } from '../common/intefaces'

export const fetchMessages = createAsyncThunk('fetchMessages', async () => {
  const res = await fetch('/messages')
  return res.json()
})

const manageMessagesSlice = createSlice({
  name: 'manageMessages',
  initialState: {
    messages: [],
    selectedMessage: {},
  },
  reducers: {
    // @ts-ignore
    setMessages: (state: IState, { payload }): void => {
      state.messages = payload.messages
    },
    // @ts-ignore
    addMessage: (state: IState, { payload }): void => {
      state.messages = [...state.messages, payload]
    },
    // @ts-ignore
    selectMessage: (state: IState, { payload }): void => {
      state.selectedMessage = payload
    },
    // @ts-ignore
    unselectMessage: (state: IState): void => {
      state.selectedMessage = {}
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMessages.fulfilled, (state, { payload }) => {
      state.messages = payload
    })
    builder.addCase(fetchMessages.rejected, (state, action) => {
      state.messages = []
    })
  },
})

export const { actions, reducer } = manageMessagesSlice
