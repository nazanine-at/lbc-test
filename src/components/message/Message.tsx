import React from 'react'
import './message.css'
import { useDispatch, useSelector } from 'react-redux'
import { unselectMessage } from '../../redux/manageMessagesSlice'
import { IState } from '../../common/intefaces'

function Message() {
  const dispatch = useDispatch()
  const { subject, userId, message } = useSelector((state: IState) => state.selectedMessage)

  const unClickDelete = () => dispatch(unselectMessage())

  // userId should be unique and is mandatory
  if (!userId) {
    return null
  }

  return (
    <div className="message">
      <button onClick={() => unClickDelete()}>delete</button>
      {subject && <p>{subject}</p>}
      {userId && <p>{userId}</p>}
      {message && <p>{message}</p>}
    </div>
  )
}

export default Message
