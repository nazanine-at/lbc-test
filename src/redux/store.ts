import { configureStore } from '@reduxjs/toolkit'
import manageMessagesSlice from './manageMessagesSlice'

export const store = configureStore({ reducer: manageMessagesSlice })
