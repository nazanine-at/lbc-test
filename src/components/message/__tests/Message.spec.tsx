import React from 'react'
import { render } from '@testing-library/react'
import Message from '../Message'
import { Provider } from 'react-redux'
import { mockStore } from '../../../common/tests'
import { initialState } from '../../../redux/manageMessagesSlice'

test('<Message/>', () => {
  const store = mockStore(initialState)

  const { container } = render(
    <Provider store={store}>
      <Message />
    </Provider>
  )
  expect(container).toMatchSnapshot()
})
