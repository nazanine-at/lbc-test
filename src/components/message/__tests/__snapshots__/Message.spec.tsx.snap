// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message/> 1`] = `<div />`;
